pinc/stages.inc

```diff
@@ -11,360 +11,537 @@ include_once($relPath.'Pool.inc');
+            _("If you have several files to submit for a single project (say a text and HTML version), zip them up together first."),
+            "</p>",
+        ]
+    ));
+
+    Activities::set(new Pool(
+        'PPV',
+        _('Post-Processing Verification'),
+        [],
+        'NOREQ', // "Peer approval. Also gives F2 access."
+        '',
+        null, // access_change_callback
+        _('Once a PPer has submitted a final e-text, it needs to be checked by a PPVer before it is posted to PG.'),
+        'ppv.php',
+        _("Post-Processor"),
+        'postproofer',
+        [
+            "<p>",
+            _("In this pool, experienced volunteers verify texts that have already been Post-Processed, and mentor new Post-Processors."),
+            sprintf(
+                _("<b>Before working in this pool</b>, please make sure you read the <a href='%s'>Post-Processing Verification Guidelines</a>."),
+                "$code_url/faq/ppv.php"
+            ),
+            sprintf(
+                _("The PPV reporting form is <a href='%s'>here</a>."),
+                "$code_url/tools/post_proofers/ppv_report.php"
+            ),
+            "</p>",
+
+            "<p>",
+            sprintf(
+                _("As always, the <a href='%s'>Post-Processing Forum</a> is available for any of your questions."),
+                get_url_to_view_forum($post_processing_forum_idx)
+            ),
+            "</p>",
+        ]
+    ));
+
+    Activities::set(new Stage(
+        'SR',
+        _('Smooth Reading'),
+        [],
+        'IMMEDIATE',
+        '',
+        null, // access_change_callback
+        _('Before a final e-text is posted to PG, it can be optionally uploaded for Smooth Reading. Anyone can volunteer to Smooth Read a text, which involves reading through the text for smoothness, marking possible errors and returning it to the PPer.'),
+        null,
+        "tools/post_proofers/smooth_reading.php"
+    ));
+
+    Activities::set(new Activity(
+        'DU',
+        _('Uploading to Project Gutenberg'),
+        [],
+        'NOREQ',
+        'OOGA',
+        null // access_change_callback
+    ));
+
+    Activities::set(new Activity(
+        "P2_mentor",
+        sprintf(_("Mentoring in round %s"), 'P2'),
+        [],
+        'NOREQ',
+        '',
+        null // access_change_callback
+    ));
+
+    // now construct various access tables
+    foreach(Activities::get_all() as $activity) {
+        foreach($activity->get_access_criteria() as $key => $value) {
+            $ACCESS_CRITERIA[$key] = $value;
+        }
+        if($activity instanceof Pool) {
+            Pools::set($activity);
+        }
+        if($activity instanceof Stage) {
+            Stages::set($activity);
+        }
+        if($activity instanceof Round) {
+            $round = $activity;
+            Rounds::set($round);
+            Rounds::set_for_number($round);
```

Comment:
Yes, This came about probably because Rounds::set($round); is in the trait which all the activities use and the others are added on for the special case of the Round.
I have to admit that the way this works duplicates the hierarchical structure of the activtity-stage-round definition to some extent and it would probably be easier to create these 'tables' inside the acivity objects like the present code does. I was reluctant to do this because it somehow didn't sem like the right thing to do. Perhaps I should get over it and try doing like that.